docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+The following is an example ruleset attached to an arbitrary service definition with the ID "SmithyService" (the
+definition itself is omitted).
+
+The ruleset describes a simple if-else case: if the sole required parameter `ResourceId` contains a specific prefix,
+route this request against a special-case endpoint. Otherwise, fall back to a constant default. The `substring` function
+used in the ruleset is an example of a standard-library function used by endpoint resolvers.
+
+This particular ruleset is kept simple for the purpose of example, naturally, a ruleset can be of arbitrary complexity as
+required by the service.
+
+```json
+{
+  "parameters": {
+    "ResourceId": {
+      "type": "string",
+      "required": "true"
+    }
+  },
+  "rules": [
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
+      "conditions": [
+        {
+          "fn": "substring",
+          "argv": [
+            {"ref": "ResourceId"},
+            0,
+            5,
+            false
+          ],
+          "assign": "resourceIdPrefix"
+        },
+        {
+          "fn": "stringEquals",
+          "argv": [
+            {"ref": "resourceIdPrefix"},
+            "gov."
+          ]
+        }
+      ],
+      "endpoint": {
+        "url": "https://gov.api"
+      }
+    },
+    {
+      "documentation": "fallback to global endpoint",
+      "conditions": [],
+      "endpoint": {
+        "url": "https://global.api"
+      }
+    }
+  ]
+}
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
+in the case of the indices).
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
+    public val resourceId: String?,
+) {
+    init {
+        require(resourceId != null) { "param resourceId is required" }
+    }
+}
+
+public fun interface SmithyServiceEndpointProvider {
+    open suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint
+}
+
+public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider {
+    public suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint {
+        val resourceIdPrefix = substring(params.resourceId, 0, 5, false)
+        if (resourceIdPrefix == "gov.") {
+            // use a special endpoint for government resources
+            return Endpoint(
+                Url.parse("https://gov.api"),
+            )
+        }
+
+        // fallback to global endpoint        
+        return Endpoint(
+            Url.parse("https://global.api"),
+        )
+    }
+}
+
+public class GetResourceResolveEndpointMiddleware(
```

Comment:
this would be `internal`.

What package do each of these go to? I would assume this would go in the `internal` package?